File: examples/session.py
"""
AgentTalk — full session flow (prove control -> declare -> join -> verify -> re-verify)

No API key and no signup. The channel creator's wallet gets 10 free calls; after
that it needs credits bought with POST /api/agenttalk/buy-key. Declare and every
join are billed to the creator, and re-verify costs the creator 1 per agent.

Proof-of-control: on-chain holdings are public, so before declare, join and
re-verify the acting agent fetches a one-time challenge for { wallet, action }
and signs it with its wallet key (EIP-191). The challenge is good for one action
and expires after 120 seconds.

Wallets: by default this generates fresh, throwaway keypairs. The signature is
accepted, but a fresh wallet holds nothing, so the demo stops at the condition
gate. To complete a full session, supply two different wallets you control that
each hold at least 1 USDC on Ethereum:

    DEMO_PRIVATE_KEY_A=0xKEY_A DEMO_PRIVATE_KEY_B=0xKEY_B python session.py

(DEMO_PRIVATE_KEY is accepted for Agent A.) Keys are read only at runtime and
never leave your machine.

Install: pip install requests eth-account
Run:     python session.py
"""

from __future__ import annotations

import json
import os

import requests
from eth_account import Account
from eth_account.messages import encode_defunct

BASE_URL = "https://skyemeta.com/api/agenttalk"


def make_account(*env_names: str):
    """A wallet from the first env-supplied private key, or a fresh throwaway one."""
    pk = next((os.environ[n] for n in env_names if os.environ.get(n)), None)
    return Account.from_key(pk) if pk else Account.create()


def prove_control(account, action: str) -> str:
    """Fetch a one-time challenge, sign the message (EIP-191), return the 0x signature."""
    resp = requests.post(
        f"{BASE_URL}/challenge",
        json={"wallet": account.address, "action": action},
        headers={"Content-Type": "application/json"},
    )
    resp.raise_for_status()
    message = resp.json()["message"]
    signed = Account.sign_message(encode_defunct(text=message), account.key)
    sig = signed.signature.hex()
    return sig if sig.startswith("0x") else "0x" + sig  # eth-account may omit 0x


def declare_channel(account, conditions: list) -> requests.Response:
    """Agent A proves control, then declares conditions for a channel."""
    signature = prove_control(account, "declare")
    body = {"wallet": account.address, "signature": signature, "conditions": conditions}
    return requests.post(f"{BASE_URL}/declare", json=body, headers={"Content-Type": "application/json"})


def join_channel(channel_id: str, account) -> requests.Response:
    """Agent B proves control, then joins. Billed to the channel creator."""
    signature = prove_control(account, "join")
    return requests.post(
        f"{BASE_URL}/join",
        json={"channelId": channel_id, "wallet": account.address, "signature": signature},
        headers={"Content-Type": "application/json"},
    )


def verify_session(session_id: str) -> dict:
    """Check if a session is still valid."""
    resp = requests.get(f"{BASE_URL}/session", params={"id": session_id})
    resp.raise_for_status()
    return resp.json()


def reverify_session(session_id: str, member) -> requests.Response:
    """Re-attest every agent against current on-chain state. Must be requested by a
    session member, signing a 'reverify' challenge; the creator pays 1 credit per agent."""
    signature = prove_control(member, "reverify")
    return requests.post(
        f"{BASE_URL}/session",
        json={"action": "reverify", "sessionId": session_id,
              "wallet": member.address, "signature": signature},
        headers={"Content-Type": "application/json"},
    )


def body_of(resp: requests.Response) -> dict:
    try:
        return resp.json()
    except ValueError:
        return {}


def explain_refusal(label: str, resp: requests.Response) -> None:
    """Explain a response that did not admit the agent.

    401          the signature was missing, stale or not from this wallet: request a
                 fresh challenge and sign again
    402          the channel creator is out of free calls and credits (buy-key)
    403 pass:false  not admitted. Today this comes back both when a condition is not
                 met and when the verification service could not produce a verdict,
                 so if you expect the wallet to qualify, retry later
    502/503      verification could not be completed: retry later
    """
    status, data = resp.status_code, body_of(resp)
    if status == 403 and data.get("pass") is False:
        print(f"\nNot admitted ({label}): the signature was accepted, but the wallet was not")
        print("admitted. Either a condition is not met (expected for a fresh throwaway wallet),")
        print("or verification was unavailable. If this wallet should qualify, retry later.")
    elif status in (502, 503):
        print(f"\nVerification unavailable ({label}). Retry in a few seconds.")
    elif status == 401:
        print(f"\nProof-of-control rejected ({label}): {data.get('error')}. "
              "Request a new challenge and sign again.")
    elif status == 402:
        print(f"\nOut of credits ({label}): the channel creator must buy credits via /buy-key.")
    else:
        print(f"\n{label} failed with HTTP {status}: {data.get('error') or resp.text}")


if __name__ == "__main__":
    agent_a = make_account("DEMO_PRIVATE_KEY_A", "DEMO_PRIVATE_KEY")
    agent_b = make_account("DEMO_PRIVATE_KEY_B")

    # Condition: wallet holds >= 1 USDC on Ethereum
    conditions = [
        {
            "type": "token_balance",
            "contractAddress": "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
            "chainId": 1,
            "threshold": "1",
        }
    ]

    print(f"Agent A: {agent_a.address}")
    print(f"Agent B: {agent_b.address}")

    print("\n=== Step 1: Prove control + declare channel ===")
    declare_resp = declare_channel(agent_a, conditions)
    print(f"HTTP {declare_resp.status_code}: {declare_resp.text}")
    channel = body_of(declare_resp) if declare_resp.ok else {}
    if not channel.get("channelId"):
        explain_refusal("declare", declare_resp)
        raise SystemExit(0)

    print("\n=== Step 2: Prove control + join channel ===")
    join_resp = join_channel(channel["channelId"], agent_b)
    print(f"HTTP {join_resp.status_code}: {join_resp.text}")
    session = body_of(join_resp) if join_resp.ok else {}
    if not session.get("sessionId"):
        explain_refusal("join", join_resp)
        raise SystemExit(0)

    print("\n=== Step 3: Verify session ===")
    print(json.dumps(verify_session(session["sessionId"]), indent=2))

    print("\n=== Step 4: Re-verify (checks current on-chain state) ===")
    reverify_resp = reverify_session(session["sessionId"], agent_a)
    print(f"HTTP {reverify_resp.status_code}: {json.dumps(body_of(reverify_resp), indent=2)}")
    if reverify_resp.status_code != 200:
        explain_refusal("re-verify", reverify_resp)
    elif body_of(reverify_resp).get("ejected"):
        print("Removed on re-verify: " + ", ".join(body_of(reverify_resp)["ejected"]))

File: examples/session.js
/**
 * AgentTalk — full session flow (prove control → declare → join → verify → re-verify)
 *
 * No API key and no signup. The channel creator's wallet gets 10 free calls; after
 * that it needs credits bought with POST /api/agenttalk/buy-key. Declare and every
 * join are billed to the creator, and re-verify costs the creator 1 per agent.
 *
 * Proof-of-control: on-chain holdings are public, so before declare, join and
 * re-verify the acting agent fetches a one-time challenge for { wallet, action }
 * and signs it with its wallet key (EIP-191). The challenge is good for one action
 * and expires after 120 seconds.
 *
 * Wallets: by default this generates fresh, throwaway keypairs. The signature is
 * accepted, but a fresh wallet holds nothing, so the demo stops at the condition
 * gate. To complete a full session, supply two different wallets you control that
 * each hold at least 1 USDC on Ethereum:
 *
 *     DEMO_PRIVATE_KEY_A=0xKEY_A DEMO_PRIVATE_KEY_B=0xKEY_B node session.js
 *
 * (DEMO_PRIVATE_KEY is accepted for Agent A.) Keys are read only at runtime and
 * never leave your machine.
 *
 * Run: npm install && node session.js
 */

import { privateKeyToAccount, generatePrivateKey } from 'viem/accounts';

const BASE_URL = 'https://skyemeta.com/api/agenttalk';

// Fresh throwaway keypairs unless you supply your own funded ones via env.
const agentA = privateKeyToAccount(process.env.DEMO_PRIVATE_KEY_A || process.env.DEMO_PRIVATE_KEY || generatePrivateKey());
const agentB = privateKeyToAccount(process.env.DEMO_PRIVATE_KEY_B || generatePrivateKey());

// Prove control of `account` for `action`: fetch a one-time challenge, sign the
// returned message (EIP-191), return the signature and the challenge's nonce. Send
// both: with the nonce, nobody else's challenge requests can cancel yours.
async function proveControl(account, action) {
  const res = await fetch(`${BASE_URL}/challenge`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ wallet: account.address, action }),
  });
  if (!res.ok) throw new Error(`challenge failed: ${res.status} ${await res.text()}`);
  const { message, nonce } = await res.json();
  return { signature: await account.signMessage({ message }), nonce };
}

async function postJson(path, body) {
  const res = await fetch(`${BASE_URL}${path}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  const data = await res.json().catch(() => ({}));
  return { status: res.status, data };
}

async function declareChannel(account, conditions) {
  const { signature, nonce } = await proveControl(account, 'declare');
  return postJson('/declare', { wallet: account.address, signature, nonce, conditions });
}

async function joinChannel(channelId, account) {
  // Billed to the channel creator. The joiner still proves control of its wallet.
  const { signature, nonce } = await proveControl(account, 'join');
  return postJson('/join', { channelId, wallet: account.address, signature, nonce });
}

async function verifySession(sessionId) {
  const res = await fetch(`${BASE_URL}/session?id=${encodeURIComponent(sessionId)}`);
  if (!res.ok) throw new Error(`verify failed: ${res.status}`);
  return res.json();
}

async function reverifySession(sessionId, member) {
  // Re-attests every agent against current on-chain state. It must be requested
  // by a session member, signing a 'reverify' challenge; the creator pays 1 credit
  // per agent.
  const { signature, nonce } = await proveControl(member, 'reverify');
  return postJson('/session', { action: 'reverify', sessionId, wallet: member.address, signature, nonce });
}

// Explain a response that did not admit the agent. Returns nothing; the caller stops.
//   401          the signature was missing, stale or not from this wallet: request a
//                fresh challenge and sign again
//   402          the channel creator is out of free calls and credits (buy-key)
//   400          the conditions or wallet fields were rejected as invalid (see error)
//   403 pass:false  a condition is not met: the verification service signed "not met"
//   503          no verdict could be obtained: retry later
//   A 400, 503 or 429 costs nothing; only a signed answer uses a call.
function explainRefusal(label, res) {
  const { status, data } = res;
  if (status === 403 && data.pass === false) {
    console.log(`\nNot admitted (${label}): a condition is not met (expected for a fresh throwaway wallet).`);
  } else if (status === 400) {
    console.log(`\nConditions rejected (${label}): ${data.error}. Nothing was charged.`);
  } else if (status === 502 || status === 503) {
    console.log(`\nVerification unavailable (${label}). Nothing was charged; retry in a few seconds.`);
  } else if (status === 401) {
    console.log(`\nProof-of-control rejected (${label}): ${data.error}. Request a new challenge and sign again.`);
  } else if (status === 402) {
    console.log(`\nOut of credits (${label}): the channel creator must buy credits via /buy-key.`);
  } else {
    console.log(`\n${label} failed with HTTP ${status}: ${data.error || JSON.stringify(data)}`);
  }
}

async function main() {
  // Condition: wallet holds >= 1 USDC on Ethereum
  const conditions = [
    {
      type: 'token_balance',
      contractAddress: '0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48',
      chainId: 1,
      threshold: '1',
    },
  ];

  console.log(`Agent A: ${agentA.address}`);
  console.log(`Agent B: ${agentB.address}`);

  console.log('\n=== Step 1: Prove control + declare channel ===');
  const channel = await declareChannel(agentA, conditions);
  console.log(`HTTP ${channel.status}: ${JSON.stringify(channel.data)}`);
  if (channel.status !== 200 || !channel.data.channelId) {
    explainRefusal('declare', channel);
    return;
  }

  console.log('\n=== Step 2: Prove control + join channel ===');
  const session = await joinChannel(channel.data.channelId, agentB);
  console.log(`HTTP ${session.status}: ${JSON.stringify(session.data)}`);
  if (session.status !== 200 || !session.data.sessionId) {
    explainRefusal('join', session);
    return;
  }

  console.log('\n=== Step 3: Verify session ===');
  const status = await verifySession(session.data.sessionId);
  console.log(JSON.stringify(status, null, 2));

  console.log('\n=== Step 4: Re-verify (checks current on-chain state) ===');
  const fresh = await reverifySession(session.data.sessionId, agentA);
  console.log(`HTTP ${fresh.status}: ${JSON.stringify(fresh.data, null, 2)}`);
  if (fresh.status !== 200) {
    explainRefusal('re-verify', fresh);
  } else if (fresh.data.ejected) {
    console.log(`Removed on re-verify: ${fresh.data.ejected.join(', ')}`);
  }
}

main().catch(console.error);
